core/src/main/scala/scalaz/std/Function.scala

```diff
@@ -158,6 +160,8 @@ trait FunctionInstances extends FunctionInstances0 {
   implicit def function1Contravariant[R]: Contravariant[? => R] =
     new Contravariant[? => R] {
       def contramap[A, B](r: A => R)(f: B => A) = r compose f
+      override def narrow[A, B](f: A => R)(implicit ev: B <~< A) =
+        Liskov.contra[-? => R, B, A](ev)(f)
```

Comment:
Can't we override this for other instances as well to avoid allocation? Maybe at least for `Kleisli`?

Reply:
It is safe to define it as `asInstanceOf` in the parent. There are no lawful functors on JVM where `contramap(downcast)` is not the same as `asInstanceOf[F[B]]` modulo things that are unobservable in Scalazzi.